File: NetDeck/CardDetailView.m
//
//  CardDetailView.m
//  Net Deck
//
//  Created by Gereon Steffens on 03.05.14.
//  Copyright © 2016 Gereon Steffens. All rights reserved.
//

#import "CardDetailView.h"
#import "CardImageViewPopover.h"
#import "CardImageCell.h"
#import "BrowserImageCell.h"
#import "CardImageViewCell.h"

@implementation CardDetailView

+(void) setupDetailViewFromPopover:(CardImageViewPopover *)popover card:(Card*)card
{
    CardDetailView* cdv = [[CardDetailView alloc] init];
    
    cdv.card = card;
    cdv.detailView = popover.detailView;
    cdv.cardName = popover.cardName;
    cdv.cardType = popover.cardType;
    cdv.cardText = popover.cardText;
    
    cdv.icon1 = popover.icon1;
    cdv.icon2 = popover.icon2;
    cdv.icon3 = popover.icon3;
    
    cdv.label1 = popover.label1;
    cdv.label2 = popover.label2;
    cdv.label3 = popover.label3;
    
    [cdv setupDetailView];
}

+(void) setupDetailViewFromCell:(CardImageCell *)cell card:(Card*)card
{
    CardDetailView* cdv = [[CardDetailView alloc] init];
    
    cdv.card = card;
    cdv.detailView = cell.detailView;
    cdv.cardName = cell.cardName;
    cdv.cardType = cell.cardType;
    cdv.cardText = cell.cardText;
    
    cdv.icon1 = cell.icon1;
    cdv.icon2 = cell.icon2;
    cdv.icon3 = cell.icon3;
    
    cdv.label1 = cell.label1;
    cdv.label2 = cell.label2;
    cdv.label3 = cell.label3;
    
    [cdv setupDetailView];
}

+(void) setupDetailViewFromBrowser:(BrowserImageCell *)cell card:(Card*)card
{
    CardDetailView* cdv = [[CardDetailView alloc] init];
    
    cdv.card = card;
    cdv.detailView = cell.detailView;
    cdv.cardName = cell.cardName;
    cdv.cardType = cell.cardType;
    cdv.cardText = cell.cardText;
    
    cdv.icon1 = cell.icon1;
    cdv.icon2 = cell.icon2;
    cdv.icon3 = cell.icon3;
    
    cdv.label1 = cell.label1;
    cdv.label2 = cell.label2;
    cdv.label3 = cell.label3;
    
    [cdv setupDetailView];
}

+(void) setupDetailViewFromIphone:(CardImageViewCell *)cell card:(Card *)card
{
    CardDetailView* cdv = [[CardDetailView alloc] init];
    
    cdv.card = card;
    cdv.detailView = cell.detailView;
    cdv.cardName = cell.cardName;
    cdv.cardType = cell.cardType;
    cdv.cardText = cell.cardText;
    
    cdv.icon1 = cell.icon1;
    cdv.icon2 = cell.icon2;
    cdv.icon3 = cell.icon3;
    
    cdv.label1 = cell.label1;
    cdv.label2 = cell.label2;
    cdv.label3 = cell.label3;
    
    [cdv setupDetailView];
}

-(void) setupDetailView
{
    self.detailView.hidden = NO;
    self.detailView.backgroundColor = [UIColor colorWithWhite:1 alpha:.7];
    self.detailView.layer.cornerRadius = 10;
    self.detailView.layer.masksToBounds = YES;
    
    Card* card = self.card;
    
#if DEBUG
    self.cardName.text = [NSString stringWithFormat:@"%@ (%@)", card.name, card.code];
#else
    self.cardName.text = card.name;
#endif
    
    // hack: remove padding from the text view
    // see https://stackoverflow.com/questions/746670/how-to-lose-margin-padding-in-uitextview
    self.cardText.textContainer.lineFragmentPadding = 0;
    self.cardText.textContainerInset = UIEdgeInsetsZero;
    self.cardText.attributedText = card.attributedText;
    
    NSString* factionName = [Faction name:card.faction];
    NSString* typeName = [CardType name:card.type];
    NSString* subtype = card.subtype;
    if (subtype)
    {
        self.cardType.text = [NSString stringWithFormat:@"%@ · %@: %@", factionName, typeName, card.subtype];
    }
    else
    {
        self.cardType.text = [NSString stringWithFormat:@"%@ · %@", factionName, typeName];
    }
    
    // labels from top: cost/strength/mu
    switch (card.type)
    {
        case NRCardTypeIdentity:
            self.label1.text = [@(card.minimumDecksize) stringValue];
            self.icon1.image = [ImageCache cardIcon];
            if (card.influenceLimit == -1)
            {
                self.label2.text = @"∞";
            }
            else
            {
                self.label2.text = [@(card.influenceLimit) stringValue];
            }
            self.icon2.image = [ImageCache influenceIcon];
            if (card.role == NRRoleRunner)
            {
                self.label3.text = [NSString stringWithFormat:@"%ld", (long)card.baseLink];
                self.icon3.image = [ImageCache linkIcon];
            }
            else
            {
                self.label3.text = @"";
                self.icon3.image = nil;
            }
            break;
            
        case NRCardTypeProgram:
        case NRCardTypeResource:
        case NRCardTypeEvent:
        case NRCardTypeHardware:
            self.label1.text = card.cost != -1 ? [NSString stringWithFormat:@"%ld", (long)card.cost] : @"";
            self.icon1.image = card.cost != -1 ? [ImageCache creditIcon] : nil;
            self.label2.text = card.strength != -1 ? [NSString stringWithFormat:@"%ld", (long)card.strength] : @"";
            self.icon2.image = card.strength != -1 ? [ImageCache strengthIcon] : nil;
            self.label3.text = card.mu != -1 ? [NSString stringWithFormat:@"%ld", (long)card.mu] : @"";
            self.icon3.image = card.mu != -1 ? [ImageCache muIcon] : nil;
            break;
            
        case NRCardTypeIce:
            self.label1.text = card.cost != -1 ? [NSString stringWithFormat:@"%ld", (long)card.cost] : @"";
            self.icon1.image = card.cost != -1 ? [ImageCache creditIcon] : nil;
            self.label2.text = @"";
            self.icon2.image = nil;
            self.label3.text = card.strength != -1 ? [NSString stringWithFormat:@"%ld", (long)card.strength] : @"";
            self.icon3.image = card.strength != -1 ? [ImageCache strengthIcon] : nil;
            break;
            
        case NRCardTypeAgenda:
            self.label1.text = [NSString stringWithFormat:@"%ld", (long)card.advancementCost];
            self.icon1.image = [ImageCache difficultyIcon];
            self.label2.text = @"";
            self.icon2.image = nil;
            self.label3.text = [NSString stringWithFormat:@"%ld", (long)card.agendaPoints];
            self.icon3.image = [ImageCache apIcon];
            break;
            
        case NRCardTypeAsset:
        case NRCardTypeOperation:
        case NRCardTypeUpgrade:
            self.label1.text = card.cost != -1 ? [NSString stringWithFormat:@"%ld", (long)card.cost] : @"";
            self.icon1.image = card.cost != -1 ? [ImageCache creditIcon] : nil;
            self.label2.text = @"";
            self.icon2.image = nil;
            self.label3.text = card.trash != -1 ? [NSString stringWithFormat:@"%ld", (long)card.trash] : @"";
            self.icon3.image = card.trash != -1 ? [ImageCache trashIcon] : nil;
            break;
            
        case NRCardTypeNone:
            NSAssert(NO, @"this can't happen");
            break;
    }
}

@end

File: NetDeck/CardDetailView.swift
//
//  CardDetailView.swift
//  NetDeck
//
//  Created by Gereon Steffens on 30.10.16.
//  Copyright © 2018 Gereon Steffens. All rights reserved.
//

import UIKit

protocol CardDetailDisplay {
    var detailView: UIView! { get }
    var cardName: UILabel! { get }
    var cardType: UILabel! { get }
    var cardText: UITextView! { get }
    
    var label1: UILabel! { get }
    var label2: UILabel! { get }
    var label3: UILabel! { get }
    var icon1: UIImageView! { get }
    var icon2: UIImageView! { get }
    var icon3: UIImageView! { get }
}

class CardDetailView {

    private var card: Card
    private var details: CardDetailDisplay
    
    static func setup(from details: CardDetailDisplay, card: Card) {
        let cdv = CardDetailView(details, card)
        cdv.setup()
    }
    
    init(_ details: CardDetailDisplay, _ card: Card) {
        self.details = details
        self.card = card
    }
    
    private func setup() {
        self.details.detailView.isHidden = false
        self.details.detailView.backgroundColor = UIColor(white: 1, alpha: 0.7)
        self.details.detailView.layer.cornerRadius = 10
        self.details.detailView.layer.masksToBounds = true
        
        if BuildConfig.debug {
            self.details.cardName.text = card.name + " (" + card.code + ")"
        } else {
            self.details.cardName.text = card.name
        }
        
        // hack: remove padding from the text view
        // see https://stackoverflow.com/questions/746670/how-to-lose-margin-padding-in-uitextview
        self.details.cardText.textContainer.lineFragmentPadding = 0
        self.details.cardText.textContainerInset = UIEdgeInsets.zero
        self.details.cardText.attributedText = card.attributedText
        
        let factionName = Faction.name(for: card.faction)
        let typeName = CardType.name(for: card.type)
        let subtype = card.subtype
        if subtype.count > 0 {
            self.details.cardType.text = String(format: "%@ · %@: %@", factionName, typeName, card.subtype)
        } else {
            self.details.cardType.text = String(format: "%@ · %@", factionName, typeName)
        }
            
        // labels from top: cost/strength/mu
        switch card.type {
        case .identity:
            self.details.label1.text = "\(card.minimumDecksize)"
            self.details.icon1.image = ImageCache.cardIcon
            if card.influenceLimit == -1 {
                self.details.label2.text = "∞"
            } else {
                self.details.label2.text = "\(card.influenceLimit)"
            }
            self.details.icon2.image = ImageCache.influenceIcon
            if card.role == .runner {
                self.details.label3.text = "\(card.baseLink)"
                self.details.icon3.image = ImageCache.linkIcon
            } else {
                self.details.label3.text = ""
                self.details.icon3.image = nil
            }
                
        case .program, .resource, .event, .hardware:
            let cost = card.costString
            let str = card.strengthString
            self.details.label1.text = cost // card.cost != -1 ? [NSString stringWithFormat:@"%ld", (long)card.cost] : @""
            self.details.icon1.image = cost.count > 0 ? ImageCache.creditIcon : nil
            self.details.label2.text = str
            self.details.icon2.image = str.count > 0 ? ImageCache.strengthIcon : nil
            self.details.label3.text = card.mu != -1 ? "\(card.mu)" : ""
            self.details.icon3.image = card.mu != -1 ? ImageCache.muIcon : nil
            
        case .ice:
            let cost = card.costString
            let str = card.strengthString
            self.details.label1.text = cost
            self.details.icon1.image = cost.count > 0 ? ImageCache.creditIcon : nil
            self.details.label2.text = ""
            self.details.icon2.image = nil
            self.details.label3.text = str
            self.details.icon3.image = str.count > 0 ? ImageCache.strengthIcon : nil
            
        case .agenda:
            self.details.label1.text = "\(card.advancementCost)"
            self.details.icon1.image = ImageCache.difficultyIcon
            self.details.label2.text = ""
            self.details.icon2.image = nil
            self.details.label3.text = "\(card.agendaPoints)"
            self.details.icon3.image = ImageCache.apIcon
            
        case .asset, .operation, .upgrade:
            let cost = card.costString
            self.details.label1.text = cost
            self.details.icon1.image = cost.count > 0 ? ImageCache.creditIcon : nil
            self.details.label2.text = ""
            self.details.icon2.image = nil
            self.details.label3.text = card.trash != -1 ? "\(card.trash)" : ""
            self.details.icon3.image = card.trash != -1 ? ImageCache.trashIcon : nil
            
        case .none:
            assert(false, "this can't happen")
        }
    }
}
